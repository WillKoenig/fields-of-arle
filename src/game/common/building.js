const upgradedBuilding = {
  forest: 'park',
  stall: 'depot',
  stable: 'doubleStall',
  moorNorth: 'dmoorNorth',
}

export const flipBuilding = building => {
  if (upgradedBuilding[building.type] === undefined) return building
  return {
    ...building,
    type: upgradedBuilding[building.type],
  }
}

export default { flipBuilding }
