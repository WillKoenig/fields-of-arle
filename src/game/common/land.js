const cellNotEmpty = cell => cell.type !== 'empty'
const rowNotEmpty = row => row.every(cellNotEmpty)
export const findSeaLevel = dikes => dikes.findIndex(rowNotEmpty)

export const buildDikes = times => ({ G, ctx, ...args }) => {
  const { currentPlayer } = ctx
  const dikes = G.players[currentPlayer].dikes.slice()
  const flatDikes = dikes.reduce((accum, row) => [...row, ...accum], [])
  const timesArg = Math.min(
    times,
    flatDikes
      .slice() // fucking reverse() will fucking mutate
      .reverse()
      .findIndex(cellNotEmpty)
  )
  const nextDikes = [
    ...flatDikes.slice(-timesArg),
    ...flatDikes.slice(0, -timesArg),
  ]
  const newDikes = [
    nextDikes.slice(9, 12),
    nextDikes.slice(6, 9),
    nextDikes.slice(3, 6),
    nextDikes.slice(0, 3),
  ]
  return {
    G: {
      ...G,
      players: {
        ...G.players,
        [currentPlayer]: {
          ...G.players[currentPlayer],
          dikes: newDikes,
        },
      },
    },
    ctx,
    ...args,
  }
}

export const arrangeItem = ({ G, ctx }, { src, dst }) => {
  const player1 = G.players[ctx.currentPlayer]

  // the player object minus the item
  let player2 = {}
  let removedItem = null
  if (src.type !== 'tokens') {
    removedItem = player1[src.type][src.row][src.col].contents[src.i]
    player2 = {
      ...player1,
      [src.type]: [
        ...player1[src.type].slice(0, src.row),
        [
          ...player1[src.type][src.row].slice(0, src.col),
          {
            ...player1[src.type][src.row][src.col],
            contents: [
              ...player1[src.type][src.row][src.col].contents.slice(0, src.i),
              ...player1[src.type][src.row][src.col].contents.slice(src.i + 1),
            ],
          },
          ...player1[src.type][src.row].slice(src.col + 1),
        ],
        ...player1[src.type].slice(src.row + 1),
      ],
    }
  } else {
    removedItem = player1[src.type][src.i]
    player2 = {
      ...player1,
      tokens: [
        ...player1.tokens.slice(0, src.i),
        ...player1.tokens.slice(src.i + 1),
      ],
    }
  }

  // the player object plus the item in a better spot
  const player3 = {
    ...player2,
    [dst.type]: [
      ...player2[dst.type].slice(0, dst.row),
      [
        ...player2[dst.type][dst.row].slice(0, dst.col),
        {
          ...player2[dst.type][dst.row][dst.col],
          contents: [
            ...player2[dst.type][dst.row][dst.col].contents,
            removedItem,
          ],
        },
        ...player2[dst.type][dst.row].slice(dst.col + 1),
      ],
      ...player2[dst.type].slice(dst.row + 1),
    ],
  }

  return {
    ...G,
    players: {
      ...G.players,
      [ctx.currentPlayer]: player3,
    },
  }
}

export const place = type => state => {
  const { G, ctx, ...args } = state
  const {
    args: [{ row, col }],
  } = args
  const seaLevel = findSeaLevel(G.players[ctx.currentPlayer].dikes)
  const oldLand = G.players[ctx.currentPlayer].land[row][col]
  if (oldLand.type !== 'empty' || row < seaLevel) return state
  return {
    G: {
      ...G,
      players: {
        ...G.players,
        [ctx.currentPlayer]: {
          ...G.players[ctx.currentPlayer],
          tokens: [...G.players[ctx.currentPlayer].tokens, ...oldLand.contents],
          land: [
            ...G.players[ctx.currentPlayer].land.slice(0, row),
            [
              ...G.players[ctx.currentPlayer].land[row].slice(0, col),
              {
                type,
                contents: [],
              },
              ...G.players[ctx.currentPlayer].land[row].slice(col + 1),
            ],
            ...G.players[ctx.currentPlayer].land.slice(row + 1),
          ],
        },
      },
    },
    ctx,
    ...args,
  }
}
