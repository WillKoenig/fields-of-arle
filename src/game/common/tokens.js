import destinations from '../destinations/index'

export const isDestination = token => Object.keys(destinations).includes(token)

export const isInventoryItemButNotPeat = token =>
  [
    'wood',
    'timber',
    'clay',
    'brick',
    'linen',
    'summerWear',
    'woolen',
    'winterWear',
    'leather',
    'leatherWear',
  ].includes(token)

const flippedToken = {
  clay: 'brick',
  wood: 'timber',
  linen: 'summerWear',
  woolen: 'winterWear',
  leather: 'leatherWear',
}

export const flip = token =>
  flippedToken[token] !== undefined ? flippedToken[token] : token

export default { flip }
