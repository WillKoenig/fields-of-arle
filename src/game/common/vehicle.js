import { VehicleSlots } from './barn'

export const VehicleSource = {
  handcart: 'handcartWagon',
  wagon: 'handcartWagon',
  peatBoat: 'peatBoatPlow',
  plow: 'peatBoatPlow',
  cart: 'cartHorseCart',
  horseCart: 'cartHorseCart',
  carriage: 'carriageDroshky',
  droshky: 'carriageDroshky',
}

const upgradedVehicle = {
  handcart: 'wagon',
  peatBoat: 'plow',
  cart: 'horseCart',
  carriage: 'droshky',
}

export const flipVehicle = vehicle => {
  if (upgradedVehicle[vehicle.type] === undefined) return vehicle
  const type = upgradedVehicle[vehicle.type]
  return {
    ...vehicle,
    type,
    contents: [
      ...vehicle.contents,
      ...VehicleSlots[type].slice(vehicle.contents.length),
    ],
  }
}
