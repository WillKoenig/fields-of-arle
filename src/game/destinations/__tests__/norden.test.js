import norden, { size } from '../norden'
import { countAnimals } from '../../common/player'

const ctx = { currentPlayer: 0 }
const G = {
  players: {
    0: {
      inventory: ['winterWear'],
      // hypothetical 1x1 farm
      land: [[{ type: 'empty', contents: ['horse'] }]],
      dikes: [[{ type: 'dike', contents: ['sheep'] }]],
      barn: {
        small1: { type: 'peatBoat', contents: [] },
        small2: { type: 'peatBoat', contents: [] },
        small3: { type: 'peatBoat', contents: [] },
        small4: { type: 'peatBoat', contents: [] },
        large1: null,
        large2: null,
        large3: { type: 'peatBoat', contents: [] },
      },
      goods: { food: 0 },
    },
    1: {},
  },
}

describe('norden load', () => {
  it('has a size', () => {
    expect(size).toEqual(2)
  })

  it('noop', () => {
    expect(norden([null, null, null])({ G, ctx })).toEqual({ G, ctx })
  })

  it('sells a sheep for 4 food', () => {
    const options = [['sheep'], null, null]
    const player = norden(options)({ G, ctx }).G.players['0']
    expect(countAnimals(player)).toEqual({ cattle: 0, horses: 1, sheep: 0 })
    expect(player.goods.food).toBe(4)
  })

  it('sell only one peatBoat, in large3 first, for 5 food', () => {
    const options = [null, ['peatBoat'], null]
    const player = norden(options)({ G, ctx }).G.players['0']
    expect(player.barn).toMatchSnapshot()
    expect(player.goods.food).toBe(5)
  })

  it('sells winterWear for 7 food', () => {
    const options = [null, null, ['winterWear']]
    const player = norden(options)({ G, ctx }).G.players['0']
    expect(player.inventory).not.toContain('winterWear')
    expect(player.goods.food).toBe(7)
  })
})
