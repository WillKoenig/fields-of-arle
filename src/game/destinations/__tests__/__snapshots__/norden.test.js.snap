// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`norden load sell only one peatBoat, in large3 first, for 5 food 1`] = `
Object {
  "large1": null,
  "large2": null,
  "large3": null,
  "small1": Object {
    "contents": Array [],
    "type": "peatBoat",
  },
  "small2": Object {
    "contents": Array [],
    "type": "peatBoat",
  },
  "small3": Object {
    "contents": Array [],
    "type": "peatBoat",
  },
  "small4": Object {
    "contents": Array [],
    "type": "peatBoat",
  },
}
`;
